File: LibraryMe.UI/LibraryMe/src/app/Features/Books/services/book-service.ts
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {BookDTO} from "../models/book-dto";
import {BookShortcutDTO} from "../models/book-shortcut-dto";
import {CreateBookDTO} from "../models/create-book-dto";
import {Injectable} from "@angular/core";
import {environment} from "../../../../environments/environment";
@Injectable({
  providedIn: 'root',
})
export class BookService {
    private baseUrl = 'https://localhost:7176';

    constructor(private http: HttpClient) {}

    getBookById(id: string): Observable<BookDTO> {
        return this.http.get<BookDTO>(`${environment.apiBaseUrl}/api/books/${id}`);
    }

    getBookShortcuts(pageSize: number = 5, pageNumber: number = 1, genreId:string=""): Observable<BookShortcutDTO[]> {
        return this.http.get<BookShortcutDTO[]>(`${environment.apiBaseUrl}/api/books/shortcuts?pageSize=${pageSize}&pageNumber=${pageNumber}&genreId=${genreId}`);
    }

    getBookSummaries(pageSize: number = 5, pageNumber: number = 1, genreId:string="", query:string=""): Observable<BookDTO[]> {
        return this.http.get<BookDTO[]>(`${environment.apiBaseUrl}/api/books/summaries?pageSize=${pageSize}&pageNumber=${pageNumber}&genreId=${genreId}&searchQuery=${query}`);
    }

    createBook(dto: CreateBookDTO): Observable<any> {
        return this.http.post(`${environment.apiBaseUrl}/api/books`, dto);
    }

    updateBook(id: string, dto: CreateBookDTO): Observable<BookDTO> {
        return this.http.put<BookDTO>(`${environment.apiBaseUrl}/api/books/${id}`, dto);
    }

    deleteBook(id: string): Observable<BookDTO> {
        return this.http.delete<BookDTO>(`${environment.apiBaseUrl}/api/books/${id}`);
    }
}

File: LibraryMe.API/BookLibrary.BAL/Services/Implementations/BookService.cs
using BookLibrary.BAL.Services.Interfaces;
using BookLibrary.DAL.Repositories.Interfaces;
using BookLibrary.DAL.Models.DTO;

namespace BookLibrary.BAL.Services.Implementations
{
    public class BookService : IBookService
    {
        private readonly IBookRepository _bookRepo;

        public BookService(IBookRepository bookRepo)
        {
            _bookRepo = bookRepo;
        }

        public async Task<BookDTO?> GetBookById(Guid id)
        {
            return await _bookRepo.GetBookById(id);
        }

        public async Task<List<BookShortcutDTO>> GetBookShortcutsAsync(int pageSize = 5, int pageNumber = 1, string genreId = null)
        {
            return await _bookRepo.GetBookShortcutsAsync(pageSize, pageNumber, genreId);
        }

        public async Task<List<BookDTO>> GetBookSummariesAsync(int pageSize = 5, int pageNumber = 1, string genreId = null, string searchQuery = "")
        {
            return await _bookRepo.GetBookSummariesAsync(pageSize, pageNumber, genreId, searchQuery);
        }

        public async Task<Guid> CreateBook(CreateBookDTO dto)
        {
            return await _bookRepo.CreateBook(dto);
        }

        public async Task<BookDTO?> UpdateBook(CreateBookDTO dto, Guid id)
        {
            return await _bookRepo.UpdateBook(dto, id);
        }

        public async Task<BookDTO?> DeleteBook(Guid id)
        {
            return await _bookRepo.DeleteBook(id);
        }
    }
}
